File: javascript/src/exercises.js
import crypto from "crypto";

import fetch from "node-fetch";
globalThis.fetch = fetch;

export function change(p1) {
  let change = [0, 0, 0, 0];

  if (p1 < 0) {
    throw new RangeError("Value cannot be negative!");
  }

  if (p1 >= 25) {
    change[0] = Math.floor(p1 / 25);
    p1 = p1 % 25;
  }

  while (p1 >= 10) {
    change[1] += 1;
    p1 = p1 - 10;
  }

  while (p1 >= 5) {
    change[2] += 1;
    p1 = p1 - 5;
  }

  change[3] = p1;

  return change;
}

export function stretched(p1) {
  p1 = p1.replace(/\s/g, "");
  //talked to julian for regular expression
  let p2 = Array.from(p1);

  p2 = p2.map((letter, index) => letter.repeat(index + 1));
  let finalWord = p2.join("");

  return finalWord;
}

export function powers(base, limit, func) {
  let pow = 0;
  let ans = 0;
  while (base ** pow <= limit) {
    ans = base ** pow;
    func(ans);
    pow++;
  }
}

export function* powersGenerator(num, max) {
  let power = 1;
  while (power <= max) {
    yield power;
    power *= num;
  }
}
//from notes

export function say(word) {
  if (word === undefined) return "";
  else
    return (say2) => {
      if (say2 === undefined) return word;
      else {
        return say(word + " " + say2);
      }
    };
}

//console.log(say("hi")());

export function makeCryptoFunctions(forKey, using, withIV) {
  crypto.createCipheriv("hex");
}

export function topTenScorers(inList) {
  let leaderboard = Object.entries(inList)
    .flatMap((teamPlayers) =>
      teamPlayers[1].map((playerArray) => [...playerArray, teamPlayers[0]])
    )
    .filter((stats) => stats[1] >= 15)
    .map((postFilter) => {
      return [postFilter[0], postFilter[2] / postFilter[1], postFilter[3]];
    })
    .map((playerStats) => {
      return {
        name: playerStats[0],
        ppg: playerStats[1],
        team: playerStats[2],
      };
    })
    .sort((a, b) => {
      return b.ppg - a.ppg;
    })
    .splice(0, 10);

  return leaderboard;
}

export async function pokemonInfo(pokemon) {
  let name = pokemon.toString();
  const response = await fetch(
    "https://pokeapi.co/api/v2/pokemon/" + name.toLowerCase()
  );
  const data = await response.json();

  return {
    id: data.id,
    name: data.name,
    weight: data.weight,
  };
}

export class Quaternion {
  constructor(a, b, c, d) {
    this.co1 = a;
    this.co2 = b;
    this.co3 = c;
    this.co4 = d;
  }

  plus(q) {
    let answer = new Quaternion(
      this.co1 + q.co1,
      this.co2 + q.co2,
      this.co3 + q.co3,
      this.co4 + q.co4
    );
    return answer;
  }

  times(q) {
    let ans1 =
      this.co1 * q.co1 - this.co2 * q.co2 - this.co3 * q.co3 - this.co4 * q.co4;
    let ans2 =
      this.co1 * q.co2 + this.co2 * q.co1 + this.co3 * q.co4 - this.co4 * q.co3;
    let ans3 =
      this.co1 * q.co3 - this.co2 * q.co4 + this.co3 * q.co1 + this.co4 * q.co2;
    let ans4 =
      this.co1 * q.co4 + this.co2 * q.co3 - this.co3 * q.co2 + this.co4 * q.co1;
    return new Quaternion(ans1, ans2, ans3, ans4);
  }

  coefficients() {
    let answer = [this.co1, this.co2, this.co3, this.co4];
    return answer;
  }
}

File: python/exercises.py
import math
import array
from cryptography.fernet import Fernet

def change (amt):
    Q = 25
    D = 10
    N = 5
    num_Q = 0
    num_D = 0
    num_N = 0
    if type(amt) != int:
        raise TypeError ("Argument must be of type int.")
    if amt < 0:
        raise ValueError ("amount cannot be negative")
    if amt >= Q:
        num_Q = math.floor(amt / Q)
        amt = amt % Q  
    if amt >= D:
        num_D = math.floor(amt / D)
        amt = amt % D
    if amt >= N:
        num_N = math.floor(amt / N)
        amt = amt % N
    return tuple([num_Q, num_D, num_N, amt])

def stretched (phrase):
    phrase = phrase.replace(" ", "")
    phrase = phrase.replace("\n", "")
    phrase = phrase.replace("\t", "")
    letters_list = list(phrase)
    index = list(range(1,len(letters_list)+1))
    letters_list = list(map(lambda x, y : x * y, letters_list, index))
    return ''.join(letters_list)

def powers (**kwargs):
    exp = 0
    pow =  kwargs['base'] ** exp
    while pow <= kwargs['limit']:
        yield(pow)
        exp += 1
        pow = kwargs['base'] ** exp

def say(word = None):
  if word == None: return ''
  else:
    return lambda say2 = None : word if say2 == None else say(word + " " + say2)

def find_first_then_lower(property, strings):
    for s in strings:
        if property(s):
            return s.lower()
    raise ValueError

def crypto_functions():
    key = Fernet.generate_key()
    fernet = Fernet(key)
    def encode(s):
        encoded = fernet.encrypt(s) 
        return encoded
    def decode(s):
        decoded = fernet.decrypt(s)
        return decoded
    return [encode, decode]

def top_ten_scorers (input):
    leaderboard = []
    holdBoard = []
    for x, y in input.items():
        if len(y[0]) >= 2:
            for i in y : 
                if i[1] >= 15:
                    var1 = (i[0])
                    var2 = str("{:.2f}".format(i[2]/i[1]))
                    var2Fl = ("{:.2f}".format(i[2]/i[1]))
                    var3 = x
                    finalList = [var1, var2Fl, var3]

                    def sortSecond(val):
                        return val[1]

                    holdBoard.append(finalList)
                    holdBoard.sort(key = sortSecond, reverse=True)
    holdBoard2 = holdBoard[0:10]
    for i in holdBoard2:
        var1 = (i[0])
        var2 = str(i[1])
        var3 = i[2]
        finalStr = var1 + "|" + var2 + "|" + var3
        leaderboard.append(finalStr)
        
    return leaderboard

class Quaternion:
    def __init__(self, a, b, c, d):
        self.a = a
        self.b = b
        self.c = c
        self.d = d

    def __add__(self, q):
        answer = Quaternion(self.a + q.a, self.b + q.b, self.c + q.c, self.d + q.d)
        return answer


    def __mul__(self, q):
        ans1 = self.a * q.a - self.b * q.b - self.c * q.c - self.d * q.d
        ans2 = self.a * q.b + self.b * q.a + self.c * q.d - self.d * q.c
        ans3 = self.a * q.c - self.b * q.d + self.c * q.a + self.d * q.b
        ans4 = self.a * q.d + self.b * q.c - self.c * q.b + self.d * q.a
        return Quaternion(ans1, ans2, ans3, ans4)

    @property
    def coefficients(self):
        answer = (self.a, self.b, self.c, self.d)
        return answer

        
